### ISSUE TITLE
CRASH and HANG attaching on AMD processors

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#3356]
While regular DR operation seems to work fine on AMD processors, attach runs into strange problems which stem from to-be-attached threads taking on  the gs base of the master thread, causing all kinds of confusion in the signal handler for the SIGUSR2 sent to take them over:

```
$ ./api.startstop
<Starting application api.startstop (129458)>
<unable to determine lib path for cross-arch execve>
<Initial options = -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<(1+x) Handling our fault in a TRY at 0x0000000000658809>
in dr_client_main
        retakeover for cur-native thread 129458
TAKEOVER: 10 threads to take over
For master thread tid=129458, 0=>gs base 0x00000000528af000
TAKEOVER: publishing takeover records
TAKEOVER: will signal thread 129459
TAKEOVER: will signal thread 129460
TAKEOVER: will signal thread 129461
TAKEOVER: will signal thread 129462
TAKEOVER: will signal thread 129463
TAKEOVER: will signal thread 129464
TAKEOVER: will signal thread 129465
TAKEOVER: will signal thread 129466
TAKEOVER: will signal thread 129467
TAKEOVER: will signal thread 129468
For SIGUSR2 in tid=129459, 0=>gs base is 0x00000000528af000
For SIGUSR2 in tid=129467, 0=>gs base is 0x00000000554d3000
For SIGUSR2 in tid=129461, 0=>gs base is 0x0000000055512000
For SIGUSR2 in tid=129464, 0=>gs base is 0x0000000055512000
For SIGUSR2 in tid=129462, 0=>gs base is 0x00000000554a4000
For SIGUSR2 in tid=129460, 0=>gs base is 0x0000000000000000
TAKEOVER: received signal in thread 129467
TAKEOVER: received signal in thread 129464
For SIGUSR2 in tid=129466, 0=>gs base is 0x00000000554a4000
For SIGUSR2 in tid=129465, 0=>gs base is 0x0000000055512000
For SIGUSR2 in tid=129463, 0=>gs base is 0x0000000000000000
<For SIGUSR2 in tid=129459, clearing instead of 129458's; res=0>
TAKEOVER: received signal in thread 129459
TAKEOVER: received signal in thread 129466
TAKEOVER: received signal in thread 129465
TAKEOVER: received signal in thread 129463
For SIGUSR2 in tid=129468, 0=>gs base is 0x0000000000000000
TAKEOVER: received signal in thread 129461
TAKEOVER: received signal in thread 129468
TAKEOVER: received signal in thread 129460
TAKEOVER: received signal in thread 129462
<Application api.startstop (129458).  Internal Error: DynamoRIO debug check failure: dr/git/src/core/unix/os.c:1981 !is_thread_tls_initialized()
(Error occurred @14 frags)
version 7.0.17906, custom build
-stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
```

This reproduces every time.

Other experiments show that on AMD, the gs base as read by arch_prctl is a bogus non-zero value at process startup, while on Intel processors it's always 0.

In our signal handler, on receiving SIGUSR2, when dcontext is NULL or when `dcontext->owning_thread` doesn't match `get_sys_thread_id()`, setting the gs base to any non-zero value fixes the problem.  Setting it to zero does *not* fix the problem.

It seems like there's a problem where the kernel fails to clear gs base on a context switch: we set it in our master thread and it bleeds over into one or more of the threads we signal.  However, I looked at the kernel context switch code and I do not yet see how it could happen.  Plus, a simple app that sets gs base in one thread and switches to another thread does *not* exhibit the problem.  I'm still trying to nail down a minimal reproducer without DR.

**Comment by derekbruening** [comment]
I can easily reproduce the problem with a simple 2-threaded app by binding
both threads to the same cpu.  The main thread creates a 2nd thread.  The
main thread then sets its gs base to some value.  Notice how the 2nd thread
immediately picks up the main thread's new value.  The 2nd thread later
sets its own base to 0 but this does not last long as it goes back to the
main thread's base:

```
$ taskset 1 ./arch_prctl
thread 906786: arch_prctl=0 => fs base 0x1f84880
thread 906786: arch_prctl=0 => gs base 0x7fff354d7570
thread 906787: arch_prctl=0 => gs base 0x7fff354d7570
thread 906786: set arch_prctl=0 => new gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: set arch_prctl=0 => new gs base (nil)
thread 906787: arch_prctl=0 => gs base (nil)
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
thread 906787: arch_prctl=0 => gs base 0x7ffdcac5d1e0
thread 906786: running
```

The source of the problem is that on AMD writing 0 to the selector register
(%gs) does not zero the hidden base!  So the context switch code in older
kernels messes up and keeps the prior thread's base.

This is fixed in more recent kernel code, where it is labeled as
"X86_BUG_NULL_SEG             X86_BUG(10) /* Nulling a selector preserves the base */".

It was first fixed in April 2016 in v4.7-rc1~188^2~53:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3e2b68d752c9e09c40d76442aa94d3b8e421b0f1
The fix is improved in Aug 2017 in v4.14-rc1~170^2~12:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=e137a4d8f4dd2e277e355495b6b2cb241a8693c3

**Comment by derekbruening** [comment]
Analysis of the scenarios we need to solve:

+ Never-before-seen thread, on attach: in master_signal_handler() we look
  for SIGUSR2 on AMD with dcontext==NULL (to handle a swap later during
  init for attach) or with dcontext->owning_thread != get_sys_thread_id().
  By limiting to SIGUSR2 we avoid extra overhead on other signals.
  If the conditions are met we set gsbase to a non-zero bogus value.

+ A signal in a was-under-DR, now-native thread.
  This includes detach.
  We're suppressing alarms during detach IIRC which should help.
  Also, if on detach we restore it back to a non-zero bogus value it had
  pre-attach, that will eliminate any problem as well.
  Looking at the code: no, actually we're setting it to NULL in
  tls_thread_free() (i#107 is known to not be fully solved).
  Solution: set to a non-zero bogus value for AMD.

+ A new thread created under DR control: a signal arriving before we set
  its gsbase in tls init.
  Xref the old -no_safe_read_tls_init where we wrote 0 to %gs in
  emit_new_thread_dynamo_start(): but w/ the AMD bug that's not enough,
  and it still leaves a race window.
  Fortunately we already solved this for i#2089 by inheriting from the parent:
  os_clone_pre() calls os_swap_dr_tls(dcontext, true) which puts a copy of
  our TLS in place with an invalid magic field.  That will solve any
  problems on AMD.

+ One thread going native (mixed-mode): os_thread_not_under_dynamo() calls
  os_swap_context(,,DR_STATE_ALL) which just like for clone invokes
  os_swap_dr_tls() which keeps the segment base but just invalidates the
  .magic field.

So whether with asynch signals arriving in these corner cases, or
safe_read_tls_magic calls faulting, I believe we can cover all the
scenarios with just a few code changes and without adding noticeable
overhead (I was afraid of having to invoke gettid on a hot path, which the
whole safe_read_tls_magic from i#2089 is avoiding).
